### PR TITLE
GH-3106 Implemented LineBreakMode to Button

### LINKED ISSUE #3106: [Enhancement] LineBreakMode for Button
##Summary
Add LineBreakMode (as currently supported on Label) to the Button control to specify text wrapping/trimming behavior. I've created an implementation as an Effect here:-
https://github.com/inthehand/InTheHand.Forms/blob/master/InTheHand.Forms/ButtonLineBreakModeEffect.cs

##API Changes

Addition of LineBreakMode property on Button. Use existing LineBreakMode enum.

##Intended Use Case
Allows you to specify the linebreak/wrapping appearance for Buttons rather than reverting to different default platform specific appearances.

### PR DESCRIPTION
**Pull request by pictos**
### Description of Change ###

I created this one because the #7810 was very old and painful to do a rebase.

Added the LineBreakMode to Button.

### Issues Resolved ### 
- fixes #3106

### API Changes ###

Added:
 - LineBreakMode Button.LineBreakMode { get; set; } //Bindable Property

### Platforms Affected ### 

- Core/XAML (all platforms)
- iOS
- Android
- UWP

### Behavioral/Visual Changes ###
Now users can set a line break mode to texts on Button, the default value doesn't affect any user.

None

### Before/After Screenshots ### 

SampleForForms
### PR Checklist ###

- [ ] Targets the correct branch
- [ ] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/ButtonLayoutManager.cs [hunk 1]
```diff
@@ -355,5 +359,16 @@ void UpdateImage()
...
+
+		void UpdateLineBreakMode()
+		{
+			AButton view = View;
+
+			if (view == null || _element == null || _renderer?.View == null)
+				return;
+
+			view.SetLineBreakMode(_element);
+			_renderer.View.SetAllCaps(_element.TextTransform == TextTransform.Default);
```
Comment (pictos) [inline Xamarin.Forms.Platform.Android/ButtonLayoutManager.cs:371]: @samhouts I used the same approach used [here](https://github.com/xamarin/Xamarin.Forms/blob/43d773d27c08ffdd43802c64676a6a670cabb7cf/Xamarin.Forms.Platform.Android/ButtonLayoutManager.cs#L281), to preserve the text as AllCaps.

File: Xamarin.Forms.Core/Button.cs [hunk 2]
```diff
@@ -69,6 +72,11 @@ Thickness IPaddingElement.PaddingDefaultValueCreator()
 		{
 			return default(Thickness);
 		}
+		public LineBreakMode LineBreakMode
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Button.cs:75]:
```suggestion

		public LineBreakMode LineBreakMode
```

**Comment by riverar** [comment]
When resizing a UWP app window using `MiddleTruncation` or `HeadTruncation`, the entire string is truncated, either to the first character or first available space (see screenshot).

Test cases
```
<Button Text="HeadTruncationHeadTruncationHeadTruncationHeadTruncationHeadTruncation" LineBreakMode="HeadTruncation"/>
<Button Text="MiddleTruncationMiddleTruncationMiddleTruncationMiddleTruncationMiddleTruncation" LineBreakMode="MiddleTruncation"/>
<Button Text="TailTruncationTailTruncationTailTruncationTailTruncationTailTruncationTailTruncation" LineBreakMode="TailTruncation"/>
```

image

File: Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+				case LineBreakMode.WordWrap:
+					textBlock.TextTrimming = TextTrimming.None;
+					textBlock.TextWrapping = TextWrapping.Wrap;
+					break;
+				case LineBreakMode.CharacterWrap:
+					textBlock.TextTrimming = TextTrimming.WordEllipsis;
+					textBlock.TextWrapping = TextWrapping.Wrap;
+					break;
+				case LineBreakMode.HeadTruncation:
+					// TODO: This truncates at the end.
```
Comment (riverar) [inline Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs:29]: Nit: Remove TODO

File: Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+					textBlock.TextTrimming = TextTrimming.WordEllipsis;
+					DetermineTruncatedTextWrapping(textBlock);
+					break;
+				case LineBreakMode.TailTruncation:
+					textBlock.TextTrimming = TextTrimming.CharacterEllipsis;
+					DetermineTruncatedTextWrapping(textBlock);
+					break;
+				case LineBreakMode.MiddleTruncation:
+					// TODO: This truncates at the end.
```
Comment (riverar) [inline Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs:38]: Nit: Remove TODO

**Comment by samhouts** [comment]
Yep, I agree, that UWP bug seems a bit nasty. @pictos Wanna take a look at that?

File: Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs (hunk 3 above)
Comment (pictos) [reply Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs:29]: @riverar, as you can see [here](https://github.com/xamarin/Xamarin.Forms/pull/11147/files#diff-2d9dd75388745f6bfedf2b80ea8d4c76L295) those comments exist before my change, I just moved all the code around. I don't think this is a good idea to remove it.

**Comment by pictos** [comment]
@samhouts, this bug isn't related to this PR, you can see it here #5284 for UWP (we have some know-bugs for Android also). I can try to solve this bug, but I think it would be better to address other PR for it. Does that make sense to you?

File: Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs (hunk 4 above)
Comment (pictos) [reply Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs:38]: @riverar, as you can see here those comments exist before my change, I just moved all the code around. I don't think this is a good idea to remove it.

**Comment by samhouts** [comment]
@pictos Very good! Thanks for clarifying!

File: Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs (hunk 3 above)
Comment (samhouts) [reply Xamarin.Forms.Platform.UAP/Extensions/TextBlockExtensions.cs:29]: .....and I think I may have added those myself 😅

**Comment by samhouts** [comment]
@pictos It looks like this may have broken this test:
image

Are you able to run that locally?

**Comment by samhouts** [comment]
More information about the failing test: Android Fast Renderers only

```
System.Exception : Error while performing WaitForElement(Marked("Page1"), "Timed out waiting for element...", null, null, null)
----> System.TimeoutException : Timed out waiting for element...
```

**Comment by riverar** [comment]
image

Waiting for Microsoft resolution on this one.

**Comment by Juansero29** [comment]
Why is the API for LineBreakMode still not included in `Xamarin.Forms 5.0.0.2012`? When is this being included?

**Comment by riverar** [comment]
@samhouts Heads up ^

**Comment by pictos** [comment]
cc @Redth

**Comment by toddanglin** [comment]
@pictos @Redth @samhouts Bump. Can you add any insight in to the timing this enhancement will be available in Xamarin Forms? Still does not appear to be available as of 5.0.0.2291.

**Comment by Pastajello** [comment]
This thing is implemented for 1,5 years but still not in the builds of XF. Unbelievable.

**Comment by riverar** [comment]
@samhouts Seems this change never made it out, are you tracking this?

**Comment by pictos** [comment]
cc: @jfversluis

**Comment by jfversluis** [comment]
Seems this was merged into main while we are releasing from 5.0.0 for a while. But this happened while I was not on the team so I can't give you the full context. Unfortunately, we don't really have a way for this to add this to Forms anymore as we're in maintanance mode and don't want to add any new APIs unless we really need to to support new platform OS features :(

If this is something you still want, it should go in .NET MAUI now.

**Comment by pictos** [comment]
Thanks for the reply

**Pull request merged**